File: libs/__tests__/sessions.test.ts
import { afterEach, beforeEach, describe, it, expect, vi } from "vitest";
import {
  pathToSlug,
  tailJsonl,
  tailJsonlBefore,
  listSessions,
  __resetScanHeadCacheForTests,
} from "../sessions";
import { join } from "node:path";
import { mkdirSync, mkdtempSync, rmSync, writeFileSync, utimesSync, statSync, readFileSync } from "node:fs";
import { tmpdir } from "node:os";

describe("pathToSlug", () => {
  it("converts Windows drive path to Claude slug", () => {
    expect(pathToSlug("C:\\projects\\my-bridge")).toBe("C--projects-my-bridge");
  });
  it("collapses dots in folder names (matches Claude's slug)", () => {
    expect(pathToSlug("C:\\projects\\my.app.vn\\my-bridge"))
      .toBe("C--projects-my-app-vn-my-bridge");
  });
  it("converts POSIX path to Claude slug", () => {
    expect(pathToSlug("/home/u/some-app")).toBe("-home-u-some-app");
  });
});

describe("tailJsonl", () => {
  it("returns full content at offset=0 and new offset at EOF", async () => {
    const dir = mkdtempSync(join(tmpdir(), "sess-"));
    const file = join(dir, "s.jsonl");
    writeFileSync(file, `{"a":1}\n{"b":2}\n`);
    const out = await tailJsonl(file, 0);
    expect(out.lines).toEqual([{ a: 1 }, { b: 2 }]);
    expect(out.offset).toBe(16);
    expect(out.lineOffsets).toEqual([0, 8]);
  });

  it("returns only new lines since offset", async () => {
    const dir = mkdtempSync(join(tmpdir(), "sess-"));
    const file = join(dir, "s.jsonl");
    writeFileSync(file, `{"a":1}\n`);
    const first = await tailJsonl(file, 0);
    writeFileSync(file, `{"a":1}\n{"b":2}\n`);
    const second = await tailJsonl(file, first.offset);
    expect(second.lines).toEqual([{ b: 2 }]);
    expect(second.lineOffsets).toEqual([8]);
  });

  it("skips incomplete trailing lines", async () => {
    const dir = mkdtempSync(join(tmpdir(), "sess-"));
    const file = join(dir, "s.jsonl");
    writeFileSync(file, `{"a":1}\n{"b":2`);
    const out = await tailJsonl(file, 0);
    expect(out.lines).toEqual([{ a: 1 }]);
    expect(out.offset).toBe(8);
    expect(out.lineOffsets).toEqual([0]);
  });

  it("offsets stay correct when lines contain multi-byte UTF-8", async () => {
    const dir = mkdtempSync(join(tmpdir(), "sess-"));
    const file = join(dir, "s.jsonl");
    const line1 = `{"v":"Quản lý"}\n`;
    const line2 = `{"v":"OK"}\n`;
    writeFileSync(file, line1 + line2);
    const out = await tailJsonl(file, 0);
    expect(out.lines).toEqual([{ v: "Quản lý" }, { v: "OK" }]);
    expect(out.lineOffsets).toEqual([0, Buffer.byteLength(line1, "utf8")]);
    expect(out.offset).toBe(Buffer.byteLength(line1 + line2, "utf8"));
  });
});

describe("tailJsonl chunked-read parity", () => {
  function multiChunkPayload(): { content: string; lineCount: number; sizeBytes: number } {
    const lines: string[] = [];
    for (let i = 0; i < 700; i++) {
      const padding = "x".repeat(800 + (i % 256));
      lines.push(JSON.stringify({ idx: i, payload: padding }));
    }
    const content = lines.join("\n") + "\n";
    return { content, lineCount: lines.length, sizeBytes: Buffer.byteLength(content, "utf8") };
  }

  it("returns the same lines as a direct file read for a multi-chunk file", async () => {
    const dir = mkdtempSync(join(tmpdir(), "sess-chunked-"));
    const file = join(dir, "big.jsonl");
    const { content, lineCount, sizeBytes } = multiChunkPayload();
    writeFileSync(file, content);
    expect(sizeBytes).toBeGreaterThan(256 * 1024);

    const out = await tailJsonl(file, 0);
    expect(out.lines.length).toBe(lineCount);
    expect(out.offset).toBe(sizeBytes);
    expect(out.lineOffsets.length).toBe(lineCount);
    expect((out.lines[0] as { idx: number }).idx).toBe(0);
    expect((out.lines[lineCount - 1] as { idx: number }).idx).toBe(lineCount - 1);
    const direct = readFileSync(file);
    for (let i = 0; i < Math.min(20, lineCount); i++) {
      const start = out.lineOffsets[i]!;
      const end = direct.indexOf(0x0A, start);
      expect(end).toBeGreaterThan(start);
      expect(JSON.parse(direct.subarray(start, end).toString("utf8"))).toEqual(out.lines[i]);
    }
  });

  it("handles a small file under one chunk", async () => {
    const dir = mkdtempSync(join(tmpdir(), "sess-chunked-"));
    const file = join(dir, "tiny.jsonl");
    writeFileSync(file, `{"a":1}\n{"b":2}\n{"c":3}\n`);
    const out = await tailJsonl(file, 0);
    expect(out.lines).toEqual([{ a: 1 }, { b: 2 }, { c: 3 }]);
    expect(out.offset).toBe(24);
    expect(out.lineOffsets).toEqual([0, 8, 16]);
  });

  it("trailing partial line is excluded; cursor lands at last newline", async () => {
    const dir = mkdtempSync(join(tmpdir(), "sess-chunked-"));
    const file = join(dir, "partial.jsonl");
    writeFileSync(file, `{"a":1}\n{"b":2}\n{"c":noterm`);
    const out = await tailJsonl(file, 0);
    expect(out.lines).toEqual([{ a: 1 }, { b: 2 }]);
    expect(out.offset).toBe(16);
  });

  it("survives a multi-byte UTF-8 char straddling a chunk boundary", async () => {
    const dir = mkdtempSync(join(tmpdir(), "sess-chunked-"));
    const file = join(dir, "boundary.jsonl");
    const CHUNK = 256 * 1024;
    const padding1 = "x".repeat(CHUNK - 100);
    const line1 = `{"v":"${padding1}"}\n`;
    const line1Bytes = Buffer.byteLength(line1, "utf8");
    const offsetInLine2 = CHUNK - line1Bytes;
    const padding2 = "y".repeat(Math.max(0, offsetInLine2 - 6));
    const line2 = `{"v":"${padding2}ảẢý"}\n`;
    const content = line1 + line2;
    writeFileSync(file, content);
    expect(Buffer.byteLength(content, "utf8")).toBeGreaterThan(CHUNK);

    const out = await tailJsonl(file, 0);
    expect(out.lines.length).toBe(2);
    const second = out.lines[1] as { v: string };
    expect(second.v.endsWith("ảẢý")).toBe(true);
    expect(out.offset).toBe(Buffer.byteLength(content, "utf8"));
  });

  it("matches direct read when called with a non-zero starting offset", async () => {
    const dir = mkdtempSync(join(tmpdir(), "sess-chunked-"));
    const file = join(dir, "resume.jsonl");
    writeFileSync(file, `{"a":1}\n{"b":2}\n{"c":3}\n{"d":4}\n`);
    const first = await tailJsonl(file, 0);
    expect(first.lines.length).toBe(4);
    const second = await tailJsonl(file, 16);
    expect(second.lines).toEqual([{ c: 3 }, { d: 4 }]);
    expect(second.offset).toBe(32);
    expect(second.lineOffsets).toEqual([16, 24]);
  });
});

describe("tailJsonlBefore", () => {
  it("returns lines ending at beforeOffset, starting on a line boundary", async () => {
    const dir = mkdtempSync(join(tmpdir(), "sess-"));
    const file = join(dir, "s.jsonl");
    writeFileSync(file, `{"a":1}\n{"b":2}\n{"c":3}\n`);
    const out = await tailJsonlBefore(file, 24, 1024);
    expect(out.lines).toEqual([{ a: 1 }, { b: 2 }, { c: 3 }]);
    expect(out.fromOffset).toBe(0);
    expect(out.lineOffsets).toEqual([0, 8, 16]);
  });

  it("starts on a line boundary when window mid-record", async () => {
    const dir = mkdtempSync(join(tmpdir(), "sess-"));
    const file = join(dir, "s.jsonl");
    writeFileSync(file, `{"a":1}\n{"b":2}\n{"c":3}\n`);
    const out = await tailJsonlBefore(file, 24, 12);
    expect(out.lines).toEqual([{ c: 3 }]);
    expect(out.fromOffset).toBe(16);
    expect(out.lineOffsets).toEqual([16]);
  });

  it("returns empty + fromOffset=0 when beforeOffset is 0", async () => {
    const dir = mkdtempSync(join(tmpdir(), "sess-"));
    const file = join(dir, "s.jsonl");
    writeFileSync(file, `{"a":1}\n`);
    const out = await tailJsonlBefore(file, 0);
    expect(out.lines).toEqual([]);
    expect(out.fromOffset).toBe(0);
  });
});

describe("listSessions", () => {
  it("hides leaf-pointer stub files that contain only `last-prompt`", () => {
    const dir = mkdtempSync(join(tmpdir(), "sessions-"));
    const stub = join(dir, "stub.jsonl");
    const real = join(dir, "real.jsonl");
    writeFileSync(stub, `{"type":"last-prompt","lastPrompt":"hi","leafUuid":"x","sessionId":"stub"}\n`);
    writeFileSync(real, `{"type":"user","message":{"role":"user","content":"hello"}}\n`);

    const sessions = listSessions(dir);
    expect(sessions.map((s) => s.sessionId)).toEqual(["real"]);
    expect(sessions[0]?.preview).toBe("hello");
  });

  it("keeps assistant-only sessions (no user line yet) but with empty preview", () => {
    const dir = mkdtempSync(join(tmpdir(), "sessions-"));
    const file = join(dir, "asst.jsonl");
    writeFileSync(file, `{"type":"assistant","message":{"role":"assistant","content":"…"}}\n`);
    const sessions = listSessions(dir);
    expect(sessions.map((s) => s.sessionId)).toEqual(["asst"]);
    expect(sessions[0]?.preview).toBe("");
  });

  it("keeps real sessions whose first user line is past the 8 KB head (huge attachment up front)", () => {
    const dir = mkdtempSync(join(tmpdir(), "sessions-"));
    const file = join(dir, "huge-head.jsonl");
    const bigBlob = "x".repeat(64 * 1024);
    const lines = [
      `{"type":"queue-operation","op":"enqueue"}`,
      `{"type":"attachment","data":"${bigBlob}"}`,
      `{"type":"user","message":{"role":"user","content":"hello past 8KB"}}`,
    ].join("\n") + "\n";
    writeFileSync(file, lines);
    const sessions = listSessions(dir);
    expect(sessions.map((s) => s.sessionId)).toEqual(["huge-head"]);
    expect(sessions[0]?.preview).toBe("hello past 8KB");
  });
});

describe("scanSessionHead cache (via listSessions)", () => {
  beforeEach(() => {
    __resetScanHeadCacheForTests();
  });

  it("reuses the cached preview when (mtime, size) is unchanged", () => {
    const dir = mkdtempSync(join(tmpdir(), "scan-cache-"));
    const file = join(dir, "real.jsonl");
    const content1 = `{"type":"user","message":{"role":"user","content":"alpha-text"}}\n`;
    const content2 = `{"type":"user","message":{"role":"user","content":"omega-text"}}\n`;
    expect(Buffer.byteLength(content1, "utf8")).toBe(Buffer.byteLength(content2, "utf8"));
    writeFileSync(file, content1);
    const first = listSessions(dir);
    expect(first[0]?.preview).toBe("alpha-text");

    const st = statSync(file);
    writeFileSync(file, content2);
    utimesSync(file, st.atime, st.mtime);

    const second = listSessions(dir);
    expect(second[0]?.preview).toBe("alpha-text");
  });

  it("misses when mtime changes", () => {
    const dir = mkdtempSync(join(tmpdir(), "scan-cache-"));
    const file = join(dir, "real.jsonl");
    writeFileSync(file, `{"type":"user","message":{"role":"user","content":"alpha"}}\n`);
    expect(listSessions(dir)[0]?.preview).toBe("alpha");

    const st = statSync(file);
    writeFileSync(file, `{"type":"user","message":{"role":"user","content":"beta"}}\n`);
    utimesSync(file, st.atime, new Date(st.mtimeMs + 5000));
    expect(listSessions(dir)[0]?.preview).toBe("beta");
  });

  it("does not cache results for missing files (or files that fail to stat)", () => {
    const dir = mkdtempSync(join(tmpdir(), "scan-cache-"));
    expect(listSessions(dir)).toEqual([]);
    const file = join(dir, "real.jsonl");
    writeFileSync(file, `{"type":"user","message":{"role":"user","content":"now real"}}\n`);
    const out = listSessions(dir);
    expect(out.map((s) => s.sessionId)).toEqual(["real"]);
    expect(out[0]?.preview).toBe("now real");
  });
});

describe("resolveSessionFile", () => {
  let tempHome: string;
  const VALID_SID = "0123abcd-4567-89ef-cdef-0123456789ab";

  beforeEach(() => {
    tempHome = mkdtempSync(join(tmpdir(), "bridge-sessions-test-"));
    process.env.HOME = tempHome;
    process.env.USERPROFILE = tempHome;
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    vi.spyOn(require("node:os"), "homedir").mockReturnValue(tempHome);
    vi.resetModules();
  });

  afterEach(() => {
    vi.restoreAllMocks();
    try {
      rmSync(tempHome, { recursive: true, force: true });
    } catch {
    }
  });

  function mkProjectDir(repoPath: string): string {
    const dir = join(tempHome, ".claude", "projects", pathToSlug(repoPath));
    mkdirSync(dir, { recursive: true });
    return dir;
  }

  it("resolves to <projects>/<slug>/<sid>.jsonl when the project dir exists", async () => {
    const repo = "/home/u/proj-a";
    const dir = mkProjectDir(repo);
    const { resolveSessionFile } = await import("../sessions");
    const file = resolveSessionFile(repo, VALID_SID);
    expect(file).toBe(join(dir, `${VALID_SID}.jsonl`));
  });

  it("returns null when the project dir does not exist", async () => {
    const { resolveSessionFile } = await import("../sessions");
    expect(resolveSessionFile("/home/u/never-claude-here", VALID_SID)).toBeNull();
  });

  it("returns null for an invalid sessionId", async () => {
    const repo = "/home/u/proj-b";
    mkProjectDir(repo);
    const { resolveSessionFile } = await import("../sessions");
    expect(resolveSessionFile(repo, "not-a-uuid")).toBeNull();
    expect(resolveSessionFile(repo, "")).toBeNull();
    expect(resolveSessionFile(repo, "../etc")).toBeNull();
  });

  it("rejects non-string repo / session inputs", async () => {
    const { resolveSessionFile } = await import("../sessions");
    expect(resolveSessionFile(null, VALID_SID)).toBeNull();
    expect(resolveSessionFile(undefined, VALID_SID)).toBeNull();
    expect(resolveSessionFile(42, VALID_SID)).toBeNull();
    expect(resolveSessionFile("/x", null)).toBeNull();
  });

  it("rejects empty repo, oversize repo, and NUL-byte repo", async () => {
    const { resolveSessionFile } = await import("../sessions");
    expect(resolveSessionFile("", VALID_SID)).toBeNull();
    expect(resolveSessionFile("a".repeat(5000), VALID_SID)).toBeNull();
    expect(resolveSessionFile("/home/u/proj\0evil", VALID_SID)).toBeNull();
  });

  it("rejects probe-by-guess: a `repoPath` whose slug points outside the projects root", async () => {
    const { resolveSessionFile } = await import("../sessions");
    expect(resolveSessionFile("/etc", VALID_SID)).toBeNull();
    expect(resolveSessionFile("../../etc/passwd", VALID_SID)).toBeNull();
    expect(resolveSessionFile("/var/log/secret", VALID_SID)).toBeNull();
  });

  it("does NOT require the .jsonl file itself to exist (caller decides)", async () => {
    const repo = "/home/u/proj-c";
    const dir = mkProjectDir(repo);
    const { resolveSessionFile } = await import("../sessions");
    const file = resolveSessionFile(repo, VALID_SID);
    expect(file).toBe(join(dir, `${VALID_SID}.jsonl`));
  });
});

File: internal/sessions/sessions_test.go
package sessions

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"
)

func TestPathToSlug(t *testing.T) {
	cases := []struct {
		name string
		in   string
		want string
	}{
		{"converts Windows drive path to Claude slug", `C:\projects\my-bridge`, "C--projects-my-bridge"},
		{"collapses dots in folder names (matches Claude's slug)", `C:\projects\my.app.vn\my-bridge`, "C--projects-my-app-vn-my-bridge"},
		{"converts POSIX path to Claude slug", "/home/u/some-app", "-home-u-some-app"},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := PathToSlug(tc.in); got != tc.want {
				t.Errorf("PathToSlug(%q) = %q, want %q", tc.in, got, tc.want)
			}
		})
	}
}

// writeFile is a fatal-on-error helper for test setup.
func writeFile(t *testing.T, path, content string) {
	t.Helper()
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		t.Fatalf("write %s: %v", path, err)
	}
}

func sessionIDs(entries []Entry) []string {
	ids := make([]string, len(entries))
	for i, e := range entries {
		ids[i] = e.SessionID
	}
	return ids
}

func TestListSessions(t *testing.T) {
	t.Run("hides leaf-pointer stub files that contain only `last-prompt`", func(t *testing.T) {
		ResetScanHeadCacheForTests()
		dir := t.TempDir()
		writeFile(t, filepath.Join(dir, "stub.jsonl"),
			`{"type":"last-prompt","lastPrompt":"hi","leafUuid":"x","sessionId":"stub"}`+"\n")
		writeFile(t, filepath.Join(dir, "real.jsonl"),
			`{"type":"user","message":{"role":"user","content":"hello"}}`+"\n")

		got := ListSessions(dir)
		ids := sessionIDs(got)
		if len(ids) != 1 || ids[0] != "real" {
			t.Fatalf("session IDs = %v, want [real]", ids)
		}
		if got[0].Preview != "hello" {
			t.Errorf("preview = %q, want %q", got[0].Preview, "hello")
		}
	})

	t.Run("keeps assistant-only sessions (no user line yet) but with empty preview", func(t *testing.T) {
		ResetScanHeadCacheForTests()
		dir := t.TempDir()
		writeFile(t, filepath.Join(dir, "asst.jsonl"),
			`{"type":"assistant","message":{"role":"assistant","content":"…"}}`+"\n")

		got := ListSessions(dir)
		ids := sessionIDs(got)
		if len(ids) != 1 || ids[0] != "asst" {
			t.Fatalf("session IDs = %v, want [asst]", ids)
		}
		if got[0].Preview != "" {
			t.Errorf("preview = %q, want empty", got[0].Preview)
		}
	})

	t.Run("keeps real sessions whose first user line is past the 8 KB head", func(t *testing.T) {
		ResetScanHeadCacheForTests()
		dir := t.TempDir()
		bigBlob := strings.Repeat("x", 64*1024)
		content := strings.Join([]string{
			`{"type":"queue-operation","op":"enqueue"}`,
			`{"type":"attachment","data":"` + bigBlob + `"}`,
			`{"type":"user","message":{"role":"user","content":"hello past 8KB"}}`,
		}, "\n") + "\n"
		writeFile(t, filepath.Join(dir, "huge-head.jsonl"), content)

		got := ListSessions(dir)
		ids := sessionIDs(got)
		if len(ids) != 1 || ids[0] != "huge-head" {
			t.Fatalf("session IDs = %v, want [huge-head]", ids)
		}
		if got[0].Preview != "hello past 8KB" {
			t.Errorf("preview = %q, want %q", got[0].Preview, "hello past 8KB")
		}
	})
}

func TestScanSessionHeadCache(t *testing.T) {
	t.Run("reuses the cached preview when (mtime, size) is unchanged", func(t *testing.T) {
		ResetScanHeadCacheForTests()
		dir := t.TempDir()
		file := filepath.Join(dir, "real.jsonl")
		// Equal byte length so the cache key's size component doesn't shift.
		content1 := `{"type":"user","message":{"role":"user","content":"alpha-text"}}` + "\n"
		content2 := `{"type":"user","message":{"role":"user","content":"omega-text"}}` + "\n"
		if len(content1) != len(content2) {
			t.Fatalf("test bug: content1 (%d) != content2 (%d) bytes", len(content1), len(content2))
		}
		writeFile(t, file, content1)
		first := ListSessions(dir)
		if len(first) != 1 || first[0].Preview != "alpha-text" {
			t.Fatalf("first list preview = %q, want alpha-text", first[0].Preview)
		}

		st, err := os.Stat(file)
		if err != nil {
			t.Fatalf("stat: %v", err)
		}
		writeFile(t, file, content2)
		// Restore atime + mtime so the cache key collides.
		if err := os.Chtimes(file, st.ModTime(), st.ModTime()); err != nil {
			t.Fatalf("chtimes: %v", err)
		}

		second := ListSessions(dir)
		if len(second) != 1 || second[0].Preview != "alpha-text" {
			t.Errorf("second list preview = %q, want alpha-text (cached)", second[0].Preview)
		}
	})

	t.Run("misses when mtime changes", func(t *testing.T) {
		ResetScanHeadCacheForTests()
		dir := t.TempDir()
		file := filepath.Join(dir, "real.jsonl")
		writeFile(t, file, `{"type":"user","message":{"role":"user","content":"alpha"}}`+"\n")
		first := ListSessions(dir)
		if len(first) != 1 || first[0].Preview != "alpha" {
			t.Fatalf("first list preview = %q, want alpha", first[0].Preview)
		}

		st, err := os.Stat(file)
		if err != nil {
			t.Fatalf("stat: %v", err)
		}
		writeFile(t, file, `{"type":"user","message":{"role":"user","content":"beta"}}`+"\n")
		future := st.ModTime().Add(5 * time.Second)
		if err := os.Chtimes(file, future, future); err != nil {
			t.Fatalf("chtimes: %v", err)
		}

		second := ListSessions(dir)
		if len(second) != 1 || second[0].Preview != "beta" {
			t.Errorf("second list preview = %q, want beta (fresh)", second[0].Preview)
		}
	})

	t.Run("does not cache results for missing files", func(t *testing.T) {
		ResetScanHeadCacheForTests()
		dir := t.TempDir()
		if got := ListSessions(dir); len(got) != 0 {
			t.Fatalf("empty dir list = %v, want empty", got)
		}

		file := filepath.Join(dir, "real.jsonl")
		writeFile(t, file, `{"type":"user","message":{"role":"user","content":"now real"}}`+"\n")
		out := ListSessions(dir)
		ids := sessionIDs(out)
		if len(ids) != 1 || ids[0] != "real" {
			t.Fatalf("session IDs = %v, want [real]", ids)
		}
		if out[0].Preview != "now real" {
			t.Errorf("preview = %q, want %q", out[0].Preview, "now real")
		}
	})
}

// TestReadSessionCwdHandlesLongLine covers the bufio.Scanner-ceiling
// regression: prior versions used a Scanner with a 1 MB max-line buffer
// over a 16 KB head slice, which tore in two ways — (a) any session
// whose first line exceeded the scanner cap aborted with
// bufio.ErrTooLong; (b) the 16 KB head was too small for modern
// transcripts whose first line embeds a multi-KB attachment. Both
// silently broke orphan-project recovery: ReadSessionCwd returned
// ("", false) and the discovered project showed up under its slug
// instead of a real path.
//
// The fix is twofold: bufio.Reader (no per-line cap) and a 256 KB head
// window. This test exercises a 200 KB line 1 that still carries the
// cwd field — neither change alone is enough; both must hold.
func TestReadSessionCwdHandlesLongFirstLine(t *testing.T) {
	dir := t.TempDir()
	file := filepath.Join(dir, "huge-line1.jsonl")

	// Build a single ~200 KB JSON line with a real cwd field. We pad the
	// `pad` field with 'x' so the line is well past any 16 KB cap but
	// below the 256 KB head window.
	pad := strings.Repeat("x", 200*1024)
	line1 := `{"cwd":"/home/u/recovered-proj","pad":"` + pad + `"}`
	body := line1 + "\n"
	writeFile(t, file, body)

	cwd, ok := ReadSessionCwd(file)
	if !ok {
		t.Fatalf("ReadSessionCwd ok = false for 200 KB line-1, want true")
	}
	if cwd != "/home/u/recovered-proj" {
		t.Errorf("cwd = %q, want /home/u/recovered-proj", cwd)
	}
}

// TestReadSessionCwdSkipsBigLinesAndFindsCwdLater verifies that when
// the head holds several large but cwd-less lines, the scan keeps going
// and recovers the cwd from a later line (here, line 5). With the old
// Scanner, the first oversize line would have aborted the scan with
// ErrTooLong and the cwd would never have been found.
func TestReadSessionCwdSkipsBigLinesAndFindsCwdLater(t *testing.T) {
	dir := t.TempDir()
	file := filepath.Join(dir, "cwd-on-line5.jsonl")

	// Four ~30 KB lines without a cwd, then a normal line that has it.
	// Total well under 256 KB so it stays in the head window.
	bigBlob := strings.Repeat("y", 30*1024)
	bigLine := `{"type":"queue-operation","payload":"` + bigBlob + `"}`
	cwdLine := `{"type":"user","cwd":"/home/u/late-cwd","message":{"role":"user","content":"hi"}}`
	body := strings.Join([]string{
		bigLine,
		bigLine,
		bigLine,
		bigLine,
		cwdLine,
	}, "\n") + "\n"
	writeFile(t, file, body)

	cwd, ok := ReadSessionCwd(file)
	if !ok {
		t.Fatalf("ReadSessionCwd ok = false when cwd is on line 5 after big lines, want true")
	}
	if cwd != "/home/u/late-cwd" {
		t.Errorf("cwd = %q, want /home/u/late-cwd", cwd)
	}
}

func TestResolveSessionFile(t *testing.T) {
	const validSID = "0123abcd-4567-89ef-cdef-0123456789ab"

	mkProjectDir := func(t *testing.T, root, repo string) string {
		t.Helper()
		dir := filepath.Join(root, PathToSlug(repo))
		if err := os.MkdirAll(dir, 0o755); err != nil {
			t.Fatalf("mkdir %s: %v", dir, err)
		}
		return dir
	}

	t.Run("resolves to <root>/<slug>/<sid>.jsonl when the project dir exists", func(t *testing.T) {
		root := t.TempDir()
		r := &Reader{Root: root}
		repo := "/home/u/proj-a"
		dir := mkProjectDir(t, root, repo)

		got, ok := r.ResolveSessionFile(repo, validSID)
		if !ok {
			t.Fatalf("ResolveSessionFile ok = false, want true")
		}
		want := filepath.Join(dir, validSID+".jsonl")
		// Compare via filepath.Clean / Abs to normalize any case/symlink
		// drift on the Windows tmp path.
		gotAbs, _ := filepath.Abs(got)
		wantAbs, _ := filepath.Abs(want)
		if gotAbs != wantAbs {
			t.Errorf("path = %q, want %q", gotAbs, wantAbs)
		}
	})

	t.Run("returns null when the project dir does not exist", func(t *testing.T) {
		root := t.TempDir()
		r := &Reader{Root: root}
		if _, ok := r.ResolveSessionFile("/home/u/never-claude-here", validSID); ok {
			t.Errorf("ok = true for missing project dir, want false")
		}
	})

	t.Run("returns null for an invalid sessionId", func(t *testing.T) {
		root := t.TempDir()
		r := &Reader{Root: root}
		repo := "/home/u/proj-b"
		mkProjectDir(t, root, repo)
		for _, sid := range []string{"not-a-uuid", "", "../etc"} {
			if _, ok := r.ResolveSessionFile(repo, sid); ok {
				t.Errorf("ok = true for sessionID %q, want false", sid)
			}
		}
	})

	t.Run("rejects empty repo, oversize repo, and NUL-byte repo", func(t *testing.T) {
		root := t.TempDir()
		r := &Reader{Root: root}
		repos := []string{"", strings.Repeat("a", 5000), "/home/u/proj\x00evil"}
		for _, repo := range repos {
			if _, ok := r.ResolveSessionFile(repo, validSID); ok {
				t.Errorf("ok = true for repo %q, want false", repo)
			}
		}
	})

	t.Run("rejects probe-by-guess: a repoPath whose slug points outside the projects root", func(t *testing.T) {
		root := t.TempDir()
		r := &Reader{Root: root}
		for _, repo := range []string{"/etc", "../../etc/passwd", "/var/log/secret"} {
			if _, ok := r.ResolveSessionFile(repo, validSID); ok {
				t.Errorf("ok = true for unmapped repo %q, want false", repo)
			}
		}
	})

	t.Run("does NOT require the .jsonl file itself to exist (caller decides)", func(t *testing.T) {
		root := t.TempDir()
		r := &Reader{Root: root}
		repo := "/home/u/proj-c"
		dir := mkProjectDir(t, root, repo)

		got, ok := r.ResolveSessionFile(repo, validSID)
		if !ok {
			t.Fatalf("ok = false, want true (project dir exists, .jsonl absence is fine)")
		}
		want := filepath.Join(dir, validSID+".jsonl")
		gotAbs, _ := filepath.Abs(got)
		wantAbs, _ := filepath.Abs(want)
		if gotAbs != wantAbs {
			t.Errorf("path = %q, want %q", gotAbs, wantAbs)
		}
		if _, err := os.Stat(got); err == nil {
			t.Errorf("the .jsonl file unexpectedly exists at %q", got)
		}
	})
}
